File: agents/go/afl-agent/mongo_ops.go
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package aflagent

import (
	"context"
	"log"

	"github.com/google/uuid"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// MongoOps provides MongoDB operations for the AFL agent protocol.
type MongoOps struct {
	db *mongo.Database
}

// NewMongoOps creates a new MongoOps instance.
func NewMongoOps(db *mongo.Database) *MongoOps {
	return &MongoOps{db: db}
}

// ClaimTask atomically claims a pending task for processing.
// Returns nil if no task is available.
func (m *MongoOps) ClaimTask(ctx context.Context, taskNames []string, taskList string) (*TaskDocument, error) {
	collection := m.db.Collection(CollectionTasks)

	filter := bson.M{
		"state":          TaskStatePending,
		"name":           bson.M{"$in": taskNames},
		"task_list_name": taskList,
	}

	update := bson.M{
		"$set": bson.M{
			"state":   TaskStateRunning,
			"updated": NowMillis(),
		},
	}

	opts := options.FindOneAndUpdate().SetReturnDocument(options.After)

	var task TaskDocument
	err := collection.FindOneAndUpdate(ctx, filter, update, opts).Decode(&task)
	if err == mongo.ErrNoDocuments {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &task, nil
}

// ReadStepParams reads the params attribute from a step.
func (m *MongoOps) ReadStepParams(ctx context.Context, stepID string) (map[string]interface{}, error) {
	collection := m.db.Collection(CollectionSteps)

	var step StepDocument
	err := collection.FindOne(ctx, bson.M{"uuid": stepID}).Decode(&step)
	if err != nil {
		return nil, err
	}

	result := make(map[string]interface{})
	for name, attr := range step.Attributes.Params {
		result[name] = attr.Value
	}

	return result, nil
}

// WriteStepReturns writes return attributes to a step.
func (m *MongoOps) WriteStepReturns(ctx context.Context, stepID string, returns map[string]interface{}) error {
	collection := m.db.Collection(CollectionSteps)

	// Build the $set update for each return field
	setFields := bson.M{}
	for name, value := range returns {
		setFields["attributes.returns."+name] = StepAttribute{
			Name:     name,
			Value:    value,
			TypeHint: inferTypeHint(value),
		}
	}

	filter := bson.M{
		"uuid":  stepID,
		"state": StepStateEventTransmit,
	}

	update := bson.M{"$set": setFields}

	_, err := collection.UpdateOne(ctx, filter, update)
	return err
}

// UpdateStepReturns merges partial return attributes into a step.
// Unlike WriteStepReturns, this does NOT require the step to be in EVENT_TRANSMIT state,
// allowing handlers to stream partial results during execution.
func (m *MongoOps) UpdateStepReturns(ctx context.Context, stepID string, partial map[string]interface{}) error {
	collection := m.db.Collection(CollectionSteps)

	setFields := bson.M{}
	for name, value := range partial {
		setFields["attributes.returns."+name] = StepAttribute{
			Name:     name,
			Value:    value,
			TypeHint: inferTypeHint(value),
		}
	}

	filter := bson.M{"uuid": stepID}
	update := bson.M{"$set": setFields}

	_, err := collection.UpdateOne(ctx, filter, update)
	return err
}

// MarkTaskCompleted marks a task as completed.
func (m *MongoOps) MarkTaskCompleted(ctx context.Context, task *TaskDocument) error {
	collection := m.db.Collection(CollectionTasks)

	update := bson.M{
		"$set": bson.M{
			"state":   TaskStateCompleted,
			"updated": NowMillis(),
		},
	}

	_, err := collection.UpdateOne(ctx, bson.M{"uuid": task.UUID}, update)
	return err
}

// MarkTaskFailed marks a task as failed with an error message.
func (m *MongoOps) MarkTaskFailed(ctx context.Context, task *TaskDocument, errorMsg string) error {
	collection := m.db.Collection(CollectionTasks)

	update := bson.M{
		"$set": bson.M{
			"state":   TaskStateFailed,
			"updated": NowMillis(),
			"error":   bson.M{"message": errorMsg},
		},
	}

	_, err := collection.UpdateOne(ctx, bson.M{"uuid": task.UUID}, update)
	return err
}

// InsertResumeTask creates an afl:resume task for the Python RunnerService.
// If facetName is non-empty, the task name includes it for visibility (e.g. "afl:resume:ns.Facet").
func (m *MongoOps) InsertResumeTask(ctx context.Context, stepID, workflowID, taskList, facetName string) error {
	collection := m.db.Collection(CollectionTasks)

	resumeName := ResumeTaskName
	if facetName != "" {
		resumeName = ResumeTaskName + ":" + facetName
	}
	now := NowMillis()
	task := TaskDocument{
		UUID:         uuid.New().String(),
		Name:         resumeName,
		RunnerID:     "",
		WorkflowID:   workflowID,
		FlowID:       "",
		StepID:       stepID,
		State:        TaskStatePending,
		Created:      now,
		Updated:      now,
		TaskListName: taskList,
		DataType:     "resume",
		Data: map[string]interface{}{
			"step_id":     stepID,
			"workflow_id": workflowID,
		},
	}

	_, err := collection.InsertOne(ctx, task)
	return err
}

// InsertStepLog inserts a step log entry for dashboard observability.
// Best-effort: errors are logged but not returned.
func (m *MongoOps) InsertStepLog(ctx context.Context, stepID, workflowID, runnerID, facetName, source, level, message string) {
	collection := m.db.Collection(CollectionStepLogs)

	now := NowMillis()
	doc := bson.M{
		"uuid":        uuid.New().String(),
		"step_id":     stepID,
		"workflow_id": workflowID,
		"runner_id":   runnerID,
		"facet_name":  facetName,
		"source":      source,
		"level":       level,
		"message":     message,
		"details":     bson.M{},
		"time":        now,
	}

	if _, err := collection.InsertOne(ctx, doc); err != nil {
		log.Printf("Could not save step log for step %s: %v", stepID, err)
	}
}

func inferTypeHint(value interface{}) string {
	switch value.(type) {
	case bool:
		return "Boolean"
	case int, int32, int64:
		return "Long"
	case float32, float64:
		return "Double"
	case string:
		return "String"
	case []interface{}:
		return "List"
	case map[string]interface{}:
		return "Map"
	default:
		return "Any"
	}
}

File: agents/typescript/fw-agent/src/mongo-ops.ts
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Db } from "mongodb";
import { v4 as uuidv4 } from "uuid";
import {
  CollectionSteps,
  CollectionStepLogs,
  CollectionTasks,
  ResumeTaskName,
  StepStateEventTransmit,
  TaskStateCompleted,
  TaskStateFailed,
  TaskStatePending,
  TaskStateRunning,
} from "./protocol";
import {
  TaskDocument,
  StepDocument,
  StepAttribute,
  nowMillis,
  inferTypeHint,
} from "./models";

/**
 * MongoDB operations for the AFL agent protocol.
 */
export class MongoOps {
  constructor(private readonly db: Db) {}

  /**
   * Atomically claims a pending task for processing.
   * Returns null if no task is available.
   */
  async claimTask(
    taskNames: string[],
    taskList: string
  ): Promise<TaskDocument | null> {
    const collection = this.db.collection<TaskDocument>(CollectionTasks);

    const filter = {
      state: TaskStatePending,
      name: { $in: taskNames },
      task_list_name: taskList,
    };

    const update = {
      $set: {
        state: TaskStateRunning,
        updated: nowMillis(),
      },
    };

    const result = await collection.findOneAndUpdate(filter, update, {
      returnDocument: "after",
    });

    return result ?? null;
  }

  /**
   * Reads the params attribute from a step.
   */
  async readStepParams(stepId: string): Promise<Record<string, unknown>> {
    const collection = this.db.collection<StepDocument>(CollectionSteps);

    const step = await collection.findOne({ uuid: stepId });
    if (!step) {
      throw new Error(`Step not found: ${stepId}`);
    }

    const result: Record<string, unknown> = {};
    const params = step.attributes?.params;
    if (params) {
      for (const [name, attr] of Object.entries(params)) {
        result[name] = attr.value;
      }
    }

    return result;
  }

  /**
   * Writes return attributes to a step.
   */
  async writeStepReturns(
    stepId: string,
    returns: Record<string, unknown>
  ): Promise<void> {
    const collection = this.db.collection<StepDocument>(CollectionSteps);

    const setFields: Record<string, StepAttribute> = {};
    for (const [name, value] of Object.entries(returns)) {
      setFields[`attributes.returns.${name}`] = {
        name,
        value,
        type_hint: inferTypeHint(value),
      };
    }

    await collection.updateOne(
      {
        uuid: stepId,
        state: StepStateEventTransmit,
      },
      { $set: setFields }
    );
  }

  /**
   * Merges partial return attributes into a step.
   * Unlike writeStepReturns, this does NOT require the step to be in EVENT_TRANSMIT state,
   * allowing handlers to stream partial results during execution.
   */
  async updateStepReturns(
    stepId: string,
    partial: Record<string, unknown>
  ): Promise<void> {
    const collection = this.db.collection<StepDocument>(CollectionSteps);

    const setFields: Record<string, StepAttribute> = {};
    for (const [name, value] of Object.entries(partial)) {
      setFields[`attributes.returns.${name}`] = {
        name,
        value,
        type_hint: inferTypeHint(value),
      };
    }

    await collection.updateOne(
      { uuid: stepId },
      { $set: setFields }
    );
  }

  /**
   * Marks a task as completed.
   */
  async markTaskCompleted(task: TaskDocument): Promise<void> {
    const collection = this.db.collection<TaskDocument>(CollectionTasks);

    await collection.updateOne(
      { uuid: task.uuid },
      {
        $set: {
          state: TaskStateCompleted,
          updated: nowMillis(),
        },
      }
    );
  }

  /**
   * Marks a task as failed with an error message.
   */
  async markTaskFailed(task: TaskDocument, errorMsg: string): Promise<void> {
    const collection = this.db.collection<TaskDocument>(CollectionTasks);

    await collection.updateOne(
      { uuid: task.uuid },
      {
        $set: {
          state: TaskStateFailed,
          updated: nowMillis(),
          error: { message: errorMsg },
        },
      }
    );
  }

  /**
   * Inserts a step log entry for dashboard observability.
   * Best-effort: errors are caught and logged.
   */
  async insertStepLog(
    stepId: string,
    workflowId: string,
    runnerId: string,
    facetName: string,
    source: string,
    level: string,
    message: string
  ): Promise<void> {
    try {
      const collection = this.db.collection(CollectionStepLogs);

      await collection.insertOne({
        uuid: uuidv4(),
        step_id: stepId,
        workflow_id: workflowId,
        runner_id: runnerId,
        facet_name: facetName,
        source,
        level,
        message,
        details: {},
        time: nowMillis(),
      });
    } catch (err) {
      console.debug(`Could not save step log for step ${stepId}:`, err);
    }
  }

  /**
   * Inserts an afl:resume task for the Python RunnerService.
   */
  async insertResumeTask(
    stepId: string,
    workflowId: string,
    taskList: string,
    facetName: string = ""
  ): Promise<void> {
    const collection = this.db.collection<TaskDocument>(CollectionTasks);

    const now = nowMillis();
    const resumeName = facetName ? `${ResumeTaskName}:${facetName}` : ResumeTaskName;
    const task: TaskDocument = {
      uuid: uuidv4(),
      name: resumeName,
      runner_id: "",
      workflow_id: workflowId,
      flow_id: "",
      step_id: stepId,
      state: TaskStatePending,
      created: now,
      updated: now,
      task_list_name: taskList,
      data_type: "resume",
      data: {
        step_id: stepId,
        workflow_id: workflowId,
      },
    };

    await collection.insertOne(task);
  }
}
